src/mlpack/methods/ann/layer/atrous_convolution.hpp

```diff
@@ -393,6 +396,9 @@ class AtrousConvolution
 
   //! Locally-stored output parameter object.
   OutputDataType outputParameter;
+
+  //! Size of the weight matrix.
+  size_t weightSize = outSize * inSize * kernelWidth * kernelHeight;
```

Comment:
We definitely shouldn't use this approach; it's very unsafe.  In this approach `weightSize` is *only* set at the time of construction.  But if the user chooses to change any of the four values `outSize`, `inSize`, `kernelWidth`, and `kernelHeight`, then `weightSize` here will be out of date.  Further, when the object is serialized or deserialized, `weightSize` will also be incorrect.

We should instead compute the value on-demand inside of `WeightSize()`.

Reply:
Also, I think this is incorrect.  In the constructor, we set the size of `weights` like this:

```
  weights.set_size((outSize * inSize * kernelWidth * kernelHeight) + outSize,
      1);
```

Thus, shouldn't the weight size be `outSize * inSize * kernelWidth * kernelHeight + outSize`?  Let me know if I overlooked something.  That last `outSize`, I believe, accounts for the bias term (which is also part of the weight matrix).

Reply:
Actually the issue said that the size of weights for the linear layer would `inSize * outSize`, so I assumed that we have to leave out the bias term.

Reply:
Yes, sorry about that, that was my mistake. :+1:

Reply:
Hi @rcurtin , @UtR491 

I know this PR was merged but was looking at it anyway as I am also thinking of picking up a layer from the issue #2590  

Now that we're done with the intro, I'll get to the reason I commented. :P

I have a doubt, is the bias term supposed to be included in the weight size calculation?
As @UtR491 mentioned, the issue doesn't specify including bias in the calculation. 
> Actually the issue said that the size of weights for the linear layer would be inSize * outSize, so I assumed that we have to leave out the bias term.

And @rcurtin agreed with it too. 
> Yes, sorry about that, that was my mistake. 👍

But in the code, the bias term was included finally. 
`  size_t WeightSize() const
  {
    return (outSize * inSize * kernelWidth * kernelHeight) + outSize;
  }
`

Which I think might be an error that has cropped up. Even if it is not, clarification on this point would certainly help me as well as any other contributor who decides to work on the issue.

Reply:
Umm, maybe @rcurtin meant that the mistake was not to include the bias term in the issue (this was what I interpreted). Moreover, right now the `weight set(size) visitor` returns `Parameters().n_elem()`, and we have to implement the same thing in a function that does not call the entire weight matrix when called, which would include the bias term too. So I think the bias term has to be included and the implementation is correct. As you can see, he also [asked](https://github.com/mlpack/mlpack/pull/2626#discussion_r495375307) me to include the bias term in the linear layer some days later. Maybe @himanshupathak21061998 's description of the issue needs an update.

Reply:
I see. I did miss that comment you linked. In that case, I will include the bias as well. 
Thanks.

Reply:
Hey, @UtR491 update my issue description :)